docs/synapse/model_format.rst

```diff
@@ -0,0 +1,234 @@
+early if the new property already exists.  This would cause the current model version to be set and any subsequent
+functions called.::
+
+    import synapse.lib.module as s_module
+
+    def getDataModel():
+        return FooBarModule.getBaseModels()[0][1]
+
+    class FooBarModule(s_module.CoreModule):
+
+        @staticmethod
+        def getBaseModels():
+            modl = {
+                'types': (
+                    ('foo:bar', {'subof': 'str', 'doc': 'A foo bar!'}),
+                ),
+                'forms': (
+                    ('foo:bar',
+                     {'ptype': 'str'},
+                     [('duck', {'defval': 'mallard', 'ptype': 'str', 'doc': 'Duck type.'})]
+                     ),
+                ),
+            }
+            name = 'foobar'
+            return ((name, modl), )
+
+        @s_module.modelrev('foobar', 201707210101)
+        def _testRev1(self):
+            '''
+            This revision adds the 'duck' property to our foo:bar nodes with its default value.
+            '''
+            try:
+                self.core.addPropDef('foo:bar:duck', form='foo:bar', defval='mallard', ptype='str', doc='Duck value!')
+            except DupPropName:
+                # The property already exists - return early
+                return
+            # Now lets migrate existing nodes to accommodate model changes.
+            rows = []
+            tick = s_common.now()
+            for iden, p, v, t in self.core.getRowsByProp('foo:bar'):
+                rows.append((iden, 'foo:bar:duck', 'mallard', tick))
+            self.core.addRows(rows)
+
+It is highly encouraged for model developers to write unit tests for any migrations which they do, in order to ensure
+that their migration functions are working correctly.
+
+Advanced CoreModule Usage
+-------------------------
+
+The CoreModule class can be used to define model handlers can be used to not just add models, but also add event
+handlers.  For example, we can define events using the @on handler for specific actions; and currently model creation
+events are specifically plumbed.  The CoreModule has a handle to the associated Cortex as well.  An example of
+extending the previous example is shown below (minus migration functions). ::
+
+    import logging
+    import synpase.eventbus as s_eventbus
+    import synapse.lib.module as s_module
+
+    logger = logging.getLogger(__name__)
+
+    def getDataModel():
+        return FooBarModule.getBaseModels()[0][1]
+
+    class FooBarModule(s_module.CoreModule):
+
+        # Override the default initCoreModule function
+        def initCoreModule(self):
+            # Define a function used for helping out during node creation.
+            self.onFormNode('foo:knight', self.onTufoFormKnight)
+            # Calling self.revCoreModl() is required by classes which override
+            # initCoreModule and define module revisions.
+            self.revCoreModl()
+
+        @staticmethod
```

Comment:
I would not encourage the use of static method decorators for these callback functions.  For many, there's a solid chance they'll need access to self.core